samza-test/src/main/java/org/apache/samza/test/framework/TestRunner.java

```diff
@@ -267,22 +272,38 @@ public TestRunner addOutputStream(CollectionStream stream) {
     return this;
   }
 
+
   /**
    * Utility to run a test configured using TestRunner
+   *
+   * @param timeout time to wait for the high level application or low level task to finish. This timeout does not include
+   *                input stream initialization time or the assertion time over output streams. This timeout just accounts
+   *                for time that samza job takes run
+   * @throws SamzaException if Samza job fails with exception and returns UnsuccessfulFinish as the statuscode
+   * @throws TimeoutException if Samza job times out as configured by timeout param
    */
-  public void run() {
+  public void run(Duration timeout) throws TimeoutException {
     Preconditions.checkState((app == null && taskClass != null) || (app != null && taskClass == null),
         "TestRunner should run for Low Level Task api or High Level Application Api");
+    Preconditions.checkState(!timeout.isZero() && !timeout.isNegative(),
+        "Timeouts should be positive");
     final LocalApplicationRunner runner = new LocalApplicationRunner(new MapConfig(configs));
+    boolean timedOut = false;
     if (app == null) {
       runner.runTask();
-      runner.waitForFinish();
+      timedOut = !runner.waitForFinish(timeout);
     } else {
       runner.run(app);
-      runner.waitForFinish();
+      timedOut = !runner.waitForFinish(timeout);
+    }
+    if (timedOut) {
+      throw new TimeoutException("Timed out waiting for application to finish");
```

Comment:
How about doing something like `Preconditions.checkState(timedOut == false, "Samza application timed out....")`? 
By this, you don't explicitly require the caller to handle `TimeoutException`. I don't see a valid code path beyond the job timeout except failing the test case. So expecting the user to handle timeout exception or bubble it up in their test by adding the exception to the method signature seems unnecessary.